### ISSUE TITLE
engine: Assertion failure in monte_carlo_combat_matrix::scale_probabilities()

### ISSUE THREAD

**Issue by shikadiqueen** [wesnoth/wesnoth#5533]
XY#6459 reported to me on Discord some time ago that the game (version 1.14.15) would crash with an assertion failure during an AI turn in the last regular scenario of my user-made campaign After the Storm. I can't quite tell what the specific conditions leading to the crash are, and from what jyrki has told me the code seems picky enough that this could well be a 1 in a million chance.

Regardless, I've been provided with a save that allows me and should allow others to reproduce the bug by just ending the player's turn without performing any additional moves:

[AtS-3-Blood_Kolo_94.gz](https://github.com/wesnoth/wesnoth/files/5968251/AtS-3-Blood_Kolo_94.gz)

**Note:** This requires After the Storm version 0.10.11 to be installed. As of this writing, that would be the latest version on the 1.14 add-ons server. Neither the player nor I have any modifications enabled. However, Monte Carlo simulation **must** be enabled in Advanced Preferences (this is the default).

Console output:
```
wesnoth-debug: src/attack_prediction.cpp:1668: static void {anonymous}::monte_carlo_combat_matrix::scale_probabilities(const std::vector<double>&, std::vector<double>&, double, unsigned int): Assertion `std::abs(std::accumulate(target.begin(), target.end(), 0.0) - 1.0) < 0.001' failed.
```

The assertion check in question is at the end of `monte_carlo_combat_matrix::scale_probabilities()` in `src/attack_prediction.cpp` and seems to remain unchanged between 1.14 and master as of this writing. While the bug is reproducible on 1.14, I would expect it to affect master as well. However, I can't make any guarantees with regards to the game behaving properly and triggering the necessary conditions leading up to the assertion failure if whoever looks at this decides to load the save on master.

**Comment by mattsc** [comment]
So here's what happening.  There are several things coming together:

1. The AI evaluates an attack combination by two units
2. The attack by the first unit is evaluated by the "normal" attack prediction, the second by the Monte Carlo algorithm
3. On the first attack, the attacker slows, and the defender can level up, but only if it kills the attacker

If any of those is not the case, the crash does not happen.  The error actually does not originate from the MC simulation, but from the MC taking the results from the first attack and noticing a problem with them, which comes down to the probability of the defender being slowed being calculated incorrectly.

The origin of the error is [this line](https://github.com/wesnoth/wesnoth/blob/master/src/attack_prediction.cpp#L1823).  `prob_initially_healthy_touched` is already the correct probability, it does not need to be multiplied by the second factor.  And the reason for that is, I believe, the fact that the slowed probabilities are already shifted to the non-slowed plane in [combat_matrix::conditional_levelup_b()](https://github.com/wesnoth/wesnoth/blob/master/src/attack_prediction.cpp#L1198).

So that's the problem.  I now have to check how to fix this.  I don't know yet whether it only needs to be fixed for the normal evaluation, or for the MC also.  And I don't want to screw up other cases.  I'll look into that, I'm just posting it in case @jyrkive knows right away what needs to be done.

**Comment by jyrkive** [comment]
Alright, the line you highlighted is meant to calculate the probability the defender is not originally slowed but the attacker slows them.

The purpose of that second factor is to correct the probability for the case the defender kills the attacker, which heals slow.

So, you're saying that `conditional_levelup_b()` already ends up correcting for that possibility, and as a result there is double correction?

Worth noting that even if removing the second factor *is* the correct fix, it should only be done for slow, not poison. Poison doesn't have separate planes (because it doesn't affect damage in later hits).

**Comment by mattsc** [comment]
> So, you're saying that `conditional_levelup_b()` already ends up correcting for that possibility, and as a result there is double correction?

That's my understanding of what that function does.  It transfers probabilities from the slowed to the non-slowed plane when a levelup occurs.  (There are a total of four of those functions, conditional and guaranteed levelups for both attacker and defender.  The one I highlighted is just the one that applies in this case.)

> Worth noting that even if removing the second factor _is_ the correct fix, it should only be done for slow, not poison. Poison doesn't have separate planes (because it doesn't affect damage in later hits).

Yes, agreed.

**Comment by mattsc** [comment]
So, yes, I confirmed that what I said earlier is correct.  Here's the attack that causes the inconsistency that triggers the assert.

<img width="645" alt="Screen Shot 2021-03-09 at 17 10 53" src="https://user-images.githubusercontent.com/2106424/110560251-87980b80-80fa-11eb-8947-9dc584c3a6ee.png">

The attacker has slow, the defender levels up if it kills the defender.  The defender has 155/223 HP.  So you can see that it has a 2.2% chance of leveling up.  Add the 6.8% of not being hit to that, we have a 9% chance of not being slowed.  Or a 91% chance of being slowed (which you can also get from 42% + 49%).

However, the code calculates the slowed chance as 89.0%.  That is equal to 91% * (1 - 0.022), meaning that the levelup "un-slowing" is double counted, just as we said earlier.

For the record, when forcing the engine to use the MC sims, it gives the correct answer.  That's because those do not use the `calculate_probability_of_debuff()` function.

So I am quite certain that that is the fix, but I will still do some more tests, just to be sure it does not mess up something else.

**Comment by mattsc** [comment]
And there's more strange stuff going on.  In this fight:
![grunt_ghoul](https://user-images.githubusercontent.com/2106424/110715817-d016fe00-81ba-11eb-80e0-9e8ae2dbef4f.jpg)
the code claims that the chance of the grunt (attacker) being poisoned is 52.2%.  Also, look at the "chance of being unscathed" of 18.5% for the attacker.  Neither number is right, and I cannot see any (simple) way of getting them from those HP stats.

So, there more investigation to be done here.

**Comment by mattsc** [comment]
Turns out the 18.5% unscathed probability is actually correct, the word is just used differently here than I first thought.  It does not mean coming out of the fight with no damage.  It means not being hit by the opponent.

**Comment by AI0867** [comment]
Is that a useful statistic to know though?

**Comment by mattsc** [comment]
> Is that a useful statistic to know though?

I'm not sure.  Are there "things" (specials, abilities, ....) that only kick in if a unit was hit, or not?

I'm not sure how useful it would be to display the other interpretation of unscathed either, since it can quite easily be read off the bar chart.  And finally, how'd we deal with things like drain?  If a unit gets the same amount of HP from drain as it looses due to an attack, does that count as unscathed?  What if it drains 1 HP more than it takes damage?  Would we add up all probabilities of HP equal to or higher than it currently has and count that into the unscathed category?

In any case, I think this is outside the scope of this issue.  I'm trying to make sure the numbers are calculated correctly, not change the current meaning of any of them.  That could be done in a separate feature request.

**Comment by AI0867** [comment]
> > Is that a useful statistic to know though?
> 
> I'm not sure. Are there "things" (specials, abilities, ....) that only kick in if a unit was hit, or not?

There's petrify, of course, but that also ends combat, so it's not relevant. I don't know about any others where it would matter. The requirement would be that a hit inflicts something that does not clear on levelup.
 
> I'm not sure how useful it would be to display the other interpretation of unscathed either, since it can quite easily be read off the bar chart. And finally, how'd we deal with things like drain? If a unit gets the same amount of HP from drain as it looses due to an attack, does that count as unscathed? What if it drains 1 HP more than it takes damage? Would we add up all probabilities of HP equal to or higher than it currently has and count that into the unscathed category?

That would be a sensible interpretation, if that hit does not afflict it with some sort of status.

> In any case, I think this is outside the scope of this issue. I'm trying to make sure the numbers are calculated correctly, not change the current meaning of any of them. That could be done in a separate feature request.

True.

**Comment by mattsc** [comment]
And the saga continues ...  While searching the forums for a definition of unscathed, I found [this post](https://forums.wesnoth.org/viewtopic.php?p=629657#p629657).

The reported discrepancy is due to the ghost having drain, but the attack prediction uses the "exact probability calculation", which does not take drain into account.  In this example, the ghost has a number of hitpoints that means that it dies after 2 hits if it does not drain the attacker, but 3 hits if it does get an HP increase from drain.  If you give the ghost 10 HP (rather than the 9 it has in that post), the discrepancy increase to 34.8% vs. 29.4%, which is definitely not negligible any more.

Options would be to use the MC simulation for this case also (which has an overall uncertainty of 1-2%, but at least the two numbers would be consistent); or to eliminate or re-define the unscathed number; or ...

As with the previous point, this is not the cause of the error originally reported here.  I am just documenting problems as I work through the calculations.

**Comment by mattsc** [comment]
I think I'm almost there.  The shifting from the slowed to the unslowed plane indeed seems to be happening correctly for the "normal" calculation also, not just the Monte Carlo sims.  So for slowing, the debuff is unnecessary and gives the wrong result.  The solution for this is simply to remove [this conditional](https://github.com/wesnoth/wesnoth/blob/master/src/attack_prediction.cpp#L2449) and always go with what's in the `else` clause.  That fixes the assert reported in this issue (and most likely the one in #5002 as well).

However, as discussed, that cannot be done for poisoned and as shown in the grunt vs. ghoul fight above, the number returned is not correct in that case either.  That one needs a different fix.  I'm going to show the relevant numbers for the attacking grunt again here, just so that we don't have to scroll up an down all the time
<img width="286" alt="Screen Shot 2021-03-12 at 08 13 58" src="https://user-images.githubusercontent.com/2106424/110970847-fea5ed80-830e-11eb-97dc-b2b2daf6ed3f.png">
So we can see that the grunt has a 59.9% chance of being poisoned, but the code returns 52.2%.  The reason for this is that the 36% chance of leveling up contains contributions from both the grunt never being hit (14.4%, to be precise; 18.5% chance of not being hit (or unscathed) minus 4.1% of not leveling up) and from being hit and healing (the rest: 36.0% - 14.4% = 21.6%).

Fortunately, fixing this is quite simple, at least for this example.  The chance of being poisoned is simply that of being hit (100% - 18.5% = 81.5%) minus the chance of being hit and leveling up (21.6%).  That gives 59.9% and thus the correct answer.  So that means that [line 1813](https://github.com/wesnoth/wesnoth/blob/master/src/attack_prediction.cpp#L1813) should become:
`		prob_debuff += prob_already_debuffed_touched - (prob_kill_if_survive - prob_kill_untouched);`
The problem here is the last probability, `prob_kill_untouched`.  It's easy to determine in this example, because it's just "prob untouched minus prob original hitpoints", but that does not work like that any more if there is drain involved, or if the unit starts with full HP.

Did anybody read that all the way through?  😉   The bottom line is, fixing the assert is simple.  Getting the poisoned percentage right still needs a bit more work.

**Comment by mattsc** [comment]
FTR, I am planning to merge #5606 and close this issue (and #5002) in a couple hours.  I will then open a separate issue with the other problems I found in the process.

**Comment by mattsc** [comment]
Fixed via a5b714e5bd
